apps/emqx_durable_storage/src/emqx_dsx.erl

```diff
@@ -0,0 +1,384 @@
+c(Name) ->
+    put(?MODULE, Name).
+
+-spec ls() -> [{_Name, pid()}].
+ls() ->
+    MS = {{?name('$1'), '$2', '_'}, [], [{{'$1', '$2'}}]},
+    gproc:select({local, names}, [MS]).
+
+more(Name) ->
+    gen_server:call(?via(Name), #more_req{}).
+
+more() ->
+    more(c()).
+
+%%================================================================================
+%% behavior callbacks
+%%================================================================================
+
+-record(s, {
+    callback :: callback(),
+    callback_state :: _,
+    db :: emqx_ds:db(),
+    poll_timeout :: pos_integer(),
+    topic :: emqx_ds:topic_filter(),
+    start :: integer(),
+    active :: true | pos_integer(),
+    its = #{} :: #{emqx_ds:stream() => emqx_ds:iterator()},
+    %% Streams that have been fully replayed and should be ignored:
+    eos = #{} :: #{emqx_ds:stream() => _},
+    renew_streams_interval :: pos_integer(),
+    %%   Poll-related fields:
+    %% In-progress poll requests:
+    polls = #{} :: #{emqx_ds:stream() => reference()},
+    %% Queue of pollable streams:
+    pq :: queue:queue(emqx_ds:stream()),
+    %% Number of streams with currently inflight poll requests:
+    inflight = 0 :: non_neg_integer(),
+    %% Number of poll requests sent to the DS so far, or since the
+    %% last `next()' call:
+    poll_count = 0 :: non_neg_integer()
+}).
+
+init(
+    #{
+        callback := Callback,
+        active := Active,
+        topic := TF,
+        start := Start,
+        db := DB,
+        renew_streams_interval := RSI,
+        poll_timeout := PollTimeout
+    }
+) ->
+    process_flag(trap_exit, true),
+    S = #s{
+        callback = Callback,
+        db = DB,
+        topic = TF,
+        start = Start,
+        active = Active,
+        renew_streams_interval = RSI,
+        poll_timeout = PollTimeout,
+        pq = queue:new()
+    },
+    {ok, poll(renew_streams(S))}.
+
+handle_call(#more_req{}, _From, S0) ->
+    S = S0#s{poll_count = 0},
+    {reply, ok, poll(S)};
+handle_call(_Call, _From, S) ->
+    {reply, {error, unknown_call}, S}.
+
+handle_cast(_Cast, S) ->
+    {noreply, S}.
+
+handle_info(renew_streams, S) ->
+    {noreply, renew_streams(S)};
+handle_info(#poll_reply{ref = Ref, payload = poll_timeout}, S0 = #s{polls = Polls}) ->
+    unalias(Ref),
+    S = maps:fold(
+        fun(Stream, R, S1) ->
+            case R of
+                Ref ->
+                    handle_poll_timeout(Stream, S1);
+                _ ->
+                    S1
+            end
+        end,
+        S0,
+        Polls
+    ),
+    {noreply, poll(S)};
+handle_info(
+    #poll_reply{ref = Ref, userdata = Stream, payload = Payload},
+    #s{inflight = Inflight, polls = Polls0} = S0
+) ->
+    case maps:take(Stream, Polls0) of
+        {Ref, Polls} ->
+            S = handle_poll_reply(Stream, Payload, S0#s{inflight = Inflight - 1, polls = Polls}),
+            {noreply, poll(S)};
+        _ ->
+            logger:error("Stray poll reply ~p", [
```

Comment:
Is it intentional not to use SLOG and snake messages for logs?

Reply:
Yes. SLOG is declared in emqx application, this code is part of emqx_durable_storage. DS cannot depend on EMQX code.